design/thin-edge-core.md

```diff
@@ -0,0 +1,50 @@
+# Thin-edge Core Vision
+
+A typical IoT application has to address:
+
+* __sophisticated business-cases__, with applications involving several vendors, actors and users;
+* __very diverse integration needs__ over operating systems, protocols, data sources, processing tools, actuators, cloud end-points;
+* __constraint environments__ with low-resources devices, large-scale deployments, high exposure to security risks,
+  prohibitive manual operations, long-term supports;
+* __safety related requirements__ with devices operating machines in hazardous contexts.
+  
+The forces being these requirements are pushing in opposite directions.
+
+* On one side, one needs a dynamic and open setting with a large set of features that can be easily extended and combined.
+  [NodeRed](https://nodered.org/) is the archetype of such tools facilitating connections of components.
+* On the other side, one needs a minimal system specifically built for a task with cherry-picked components.
+  On the extreme side this leads to the idea of [unikernel](https://en.wikipedia.org/wiki/Unikernel) and sealed, fixed-purpose software images.
+  
+Can these two poles be reconciled?
+
+The approach of thin-edge is to ease the development of IoT applications on edge devices with a smooth transition between:
+* prototyping use-cases - when the aim is to easily explore and combine features,
+* production use-cases - when the need is to deploy a hardened specific software on a large fleet of devices,
+* as well as intermediate use-cases - when the application is build by several vendors with components written using different programming languages;
+  or when a manufacturer assembles his tools on the devices while letting open to his customers the option to add their own modules. 
+
+To achieve this goal, the foundations of Thin-edge address the hardened case first.
+* The core of thin-edge and its components are written using the *Rust* programming languages.
+* The set of components used by an application is defined at *build time*.
+* These components cooperate using a streaming, asynchronous message passing, *internal* API.
```

Comment:
> streaming

Does streaming here mean in the same way that HTTP does it? With partial messages + "there will be more later"?

Reply:
No, not as HTTP streaming. A single message is delivered or not. There is no half messages.

I mean as pub/sub protocols, where request/responses messages are sent independently. As in MQTT, web-socket, rust/go channels, kafka ...